### PR TITLE
ENT-6219 Added standalone self upgrade capability for Windows agents (3.12.x)

### PR DESCRIPTION
**Pull request by craigcomstock**
Also includes commented out code to use msiexec.bat packages module when ENT-6823 is resolved.

Since this policy is standalone several msiexec bits had to be copied here

Ticket: ENT-6219
Changelog: title
(cherry picked from commit a6cfc5725d212bd86961d6b83999b1304ad4be42)

merge together
https://github.com/cfengine/masterfiles/pull/1944
https://github.com/cfengine/core/pull/4548

### CODE REVIEW

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6213/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6213/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6213/

**Comment by craigcomstock** [comment]
Testing in the cloud.

File: controls/update_def.cf.in [hunk 1]
```diff
@@ -79,7 +79,8 @@ bundle common update_def
       "input_name_patterns"
         slist => { ".*\.cf",".*\.dat",".*\.txt", ".*\.conf", ".*\.mustache",
                    ".*\.sh", ".*\.pl", ".*\.py", ".*\.rb", ".*\.sed", ".*\.awk",
-                   "cf_promises_release_id", ".*\.json", ".*\.yaml" },
+                   "cf_promises_release_id", ".*\.json", ".*\.yaml", ".*\.csv",
```
Comment (craigcomstock) [inline controls/update_def.cf.in:82]: maybe don't include .csv in 3.12.x, TODO find source of that bit.

File: modules/packages/msiexec.bat [hunk 2]
```diff
@@ -101,7 +101,7 @@ rem Install this file if it exists
     mkdir %log_dir%
   )
   for /F "delims=" %%i in (%1) do @set basename="%%~ni"
-  REM %log_dir:"=% replaces quotes with nothing, otherwise you get two double-quotes which causes failures
+  REM %log_dir:"=% replaces quotes with nothing, otherwise you get two double-quotes which causes failures
```
Comment (craigcomstock) [inline modules/packages/msiexec.bat:104]: yah, dumb whitespace diff (line ending?)

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6230/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6230/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6230/

**Comment by craigcomstock** [comment]
@cf-bottom jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6239/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6239/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6239/

**Comment by craigcomstock** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6258/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6258/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6258/

**Comment by craigcomstock** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6272/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6272/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6272/

File: cfe_internal/update/windows_unattended_upgrade.cf [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+Taskkill /IM cf-monitord.exe /F
+Taskkill /IM cf-serverd.exe /F
+Taskkill /IM cf-agent.exe /F
+set bat="$(sys.workdir)$(const.dirsep)modules$(const.dirsep)packages$(const.dirsep)msiexec.bat"
+echo File=$(cfengine_software.local_software_dir)$(const.dirsep)$(cfengine_package_names.my_pkg) | call %bat% file-install
+REM exit 0 so that scheduled task will not keep trying a failing situation
+REM looking at package module logs of one failure should help in debugging
+exit 0
+`,
+        template_method => "inline_mustache",
```
Comment (nickanderson) [inline cfe_internal/update/windows_unattended_upgrade.cf:68]: Whoa, this threw me off. inline_mustache, but cfengine variable dereferences.

It's fine, I just expected to see mustache markup in it.

File: controls/update_def.cf.in (hunk 1 above)
Comment (nickanderson) [reply controls/update_def.cf.in:82]: https://github.com/cfengine/masterfiles/commit/dde9b9ec76820a12800575c919f00934b51b7baa

Comment (nickanderson) [reply controls/update_def.cf.in:82]: https://tracker.mender.io/browse/CFE-3425
Indicates its in 3.17 and 3.15, so yeah, I wouldn't bring the .csv back to 3.12

File: modules/packages/msiexec.bat (hunk 2 above)
Comment (nickanderson) [reply modules/packages/msiexec.bat:104]: `¯\_(ツ)_/¯`

File: standalone_self_upgrade.cf.in [hunk 4]
```diff
@@ -213,17 +233,20 @@ bundle agent cfengine_software_version
                 "aix",
                 "SUSE_11",
                 "SUSE_12",
+                "windows", # ENT-4094
```
Comment (nickanderson) [inline standalone_self_upgrade.cf.in:236]: i like all these ticket comments.

File: standalone_self_upgrade.cf.in [hunk 5]
```diff
@@ -909,6 +956,14 @@ body package_module apt_get
       query_updates_ifelapsed => "30";
 }
 
+body package_module msiexec
+# @brief msiexec package module default settings
+{
+      query_installed_ifelapsed => "10";
```
Comment (nickanderson) [inline standalone_self_upgrade.cf.in:962]: I would set this to 0 I think

**Review by nickanderson — CHANGES_REQUESTED** [review]
I think just drop the csv addition for input_name_patterns

File: standalone_self_upgrade.cf.in (hunk 5 above)
Comment (craigcomstock) [reply standalone_self_upgrade.cf.in:962]: Why 0 for 3.12.x and not for 3.15.x and master?

Comment (nickanderson) [reply standalone_self_upgrade.cf.in:962]: https://tracker.mender.io/browse/CFE-3504

Comment (nickanderson) [reply standalone_self_upgrade.cf.in:962]: https://tracker.mender.io/browse/CFE-3613

**Review by nickanderson — APPROVED** [review]

**Pull request merged**